### PR TITLE
Proposal: Configmaps populating environment variables

### PR DESCRIPTION
**Pull request by fraenkel**
Copied from https://github.com/kubernetes/kubernetes/pull/37295

Allow configmaps to populate the environment variables of containers.

### CODE REVIEW

**Comment by kad** [comment]
hm, this partially can also solve kubernetes/kubernetes#37448 empty referenced ConfigMap would be considered ok in the implementation side.

**Comment by derekwaynecarr** [comment]
FYI @pmorie @jpeeler

**Comment by fraenkel** [comment]
@thockin @dhoer
Adding the words to enhance kubectl create configmap is easy. The behavior is a bit tricky. In a simple world, I can point to a directory of env files and say `kubectl create configmap --from-file somedir`. Duplicates are where things become tricky. What order are we processing the files in the directory? Are we sorting them? Or letting the filesystem hand them in some random order?
I will also assume you can provide multiple directories to allow overrides.

**Comment by kad** [comment]
@fraenkel sorting would be expected behavior and more logical option

**Comment by dhoer** [comment]
@fraenkel @thockin For directory, I would assume whatever order it reads today, where on duplicated keys, the last one wins.  They can use multiple --from-file entries to get the order they want. Subsequent --from-file entries are will overwrite any repeating key values.

For example:
dir/a.env
```
  x=1
  y=2
```
dir/b.env
```
  y=3
  z=4
```

For `--from-file=./dir` would result in:
```
x=1
y=3
z=4
```

For `--from-file=./dir/b.env --from-file=./dir/a.env` would result in:
```
x=1
y=2
z=4
```

**Comment by fraenkel** [comment]
@dhoer Realize that today, you don't see this issue because the keys are unique per directory since it is the filename which is why I wanted to raise the awareness of this new behavior.

**Comment by pmorie** [comment]
@dhoer I am not sure your comments are about the API feature being added or some aspect of kubectl's behavior. Can you clarify?

**Comment by pmorie** [comment]
I would prefer to treat kubectl enhancements as a separate issue. My understanding was that API changes have been agreed to. If we want to change the behavior of kubectl it should be another proposal. If the API changes have been agreed to, let's get this in and get the feature implemented. Changes, in general, are much easier to land if they are scoped specifically at the proposal stage. We will definitely not be implementing API changes and kubectl changes in the same PR.

**Comment by fraenkel** [comment]
@pmorie My intent is to make sure the proposal covers all aspects including any CLI changes that would be needed to complete the proposal. One could argue the create configmap could stand on its own but the use case has been described as part of the ongoing discussion just like prefixes.

There will be as many PRs as we deem necessary. Right now I see at least 2, the one I have already started and then a separate one for the this create configmap change.

**Comment by thockin** [comment]
I was thinking `--from-env-file=filename` and only allowing one, and not a directory.

File: contributors/design-proposals/envvar-configmap.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+a container, where the `Env` takes precedence.  The introduction of ConfigMaps
+adds a third possibility. To prevent any change in behavior, the `Env` object
+will still override any environment variable introduced by a ConfigMap.
+A ConfigMap is allowed to override variables defined by services.  Variable
+references defined by an `EnvVar` struct can be resolved by values defined in
+other `EnvVar`s or those introduced by ConfigMaps and services.
+
+To prevent collisions amongst multiple ConfigMaps, each defined ConfigMap can
+have an optional associated prefix that is prepended to each key in the
+ConfigMap.  Prefixes must be a "C" identifier.
```
Comment (thockin) [inline contributors/design-proposals/envvar-configmap.md]: isn't it just ordering.

1) All automatic service envvars
2) All Envfrom blocks, in order
3) All Env blocks, in order

**Comment by thockin** [comment]
This is @pmorie's to final review.  Nits about ordering/collisions and kubectl.

**Comment by kad** [comment]
Two more comments to make this proposal usable in more wider scope:
1. if we implement that reference model for populating pod environment from config maps, it makes sense to do similar thing for Secrets via "secretRef". Same rules with exposing set of secrets via environment vars can be very useful and save a lot of time/lines of writing pod spec with multiple secretes to be exposed.

2.  If we already touching this part of API spec, it might be good also to add one thing: referenced ConfigMap might be optional. By default this option will be false, so if ConfigMap referenced and it is missing, it will lead to error. If optional parameter set to true, and referenced ConfigMap not found, it will continue normally.  Use case for such behavior: 
  -  Application developer can prepare Application deployment with referencing ConfigMap with default values. Developer also can reference optional ConfigMap with parameters that can be overridden in e.g. testing cluster, or regional cluster. If those overrides not found or not used, application will get all the rest of default values from ConfigMap of defaults, shipped with Deployment spec.
  - Same deployment in public and private cloud: *_proxy variables.. Developer can write spec that will be referencing ConfigMap with some agreed name, e.g. "cluster-proxies". In case of deployment of this application in private cluster this ConfigMap would contain http_proxy, https_proxy, no_proxy and other variables specific to this cluster. Same deployment sent to cluster in public cloud with direct internet access will be referencing non-existing optional ConfigMap, so containers will be run without pre-populated proxy variables.

makes sense ?

**Comment by fraenkel** [comment]
@kad 
1. See the last line in the proposal which refers to Secrets.
1. Why would we have different behavior for ConfigMap processing vs EnvVar processing? It too fails if the ConfigMap is not there.

I would rather add the optional feature after we see what pain people run into. We can always add features but removing something like this would be difficult.

**Comment by kad** [comment]
@fraenkel sorry, missed line about secrets, mine bad.
For second thing, I've already referenced in earlier comments kubernetes/kubernetes#37448 which talks about optional EnvVars. Scenarios that I mentioned in comments are my real world examples where I'd like to have one set of deployment yaml files for application and ability to override via optional vars or configmaps specifics of development cluster inside company firewall and on production cluster in GCP.

**Comment by thockin** [comment]
We may yet end up with "if present" configmap volumes - it would actually
simplify a whole swath of scenarios.  But we should do that and this in
tandem.

I agree that we should set this up to add secret support - when we do that
we will want a discriminator field, too.  the configmap ref should be
optional here.

**Comment by fraenkel** [comment]
ConfigMapRef is already optional and the EnvFrom can easily be extended to support a SecretRef.

**Comment by fraenkel** [comment]
Doing some investigation on the --from-env-file, I noticed that the current `create configmap` support does not allow duplicate keys. I realize that we can have this behave differently but reading the use cases, people expect to be able to override yet you cannot do that today with what is currently there.

**Comment by pmorie** [comment]
> ConfigMapRef is already optional and the EnvFrom can easily be extended to support a SecretRef.

Yes, `EnvFrom` is a union-type with only a single type for now.

I also have another proposal that adds `ServiceRef` to `EnvFrom` - kubernetes/kubernetes#37612

So, I personally consider the kubectl discussion to make this entire thread far more difficult to follow.  I would **strongly** prefer to start another feature request for kubectl changes, because prior experience with such changes suggest that the discussion will be lengthy, and this proposal, IMO, is now in a mergeable state.

**Comment by thockin** [comment]
I don't understand the concern here, --from-env-file doesn't need duplicate
keys?

**Comment by fraenkel** [comment]
From listening to what people what to do and how the current configmap options work, it is best to discuss our options in a separate proposal. The current --from-file and --from-literal both take multiples. --from-file also deals with directories. Both do not allow duplicates or overrides.
Adding a --from-env-file which is singular and cannot be used in conjunction with the others while doable just seems a bit wrong to me.
We can proceed with what this current proposal is stating and have a further discussion on how configmaps should or should not be created. I would suggest we start with the use-cases since they will drive out the inconsistencies we have and what our options are going forward.

**Comment by AdoHe** [comment]
@kubernetes/sig-cli

File: contributors/design-proposals/envvar-configmap.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+## Proposed Design
+
+Containers can specify a set of sources that are consumed as environment
+variables. One such source is a ConfigMap.
+Each key defined in the ConfigMap's `Data` object must be a "C" identifier. If
+an invalid key is present, the container will fail to start.
+
+Environment variables defined by a `Container` are processed in a specific
+order.  The processing order is as follows:
+1. All automatic service environment variables
```
Comment (pmorie) [inline contributors/design-proposals/envvar-configmap.md]: This does not render as a markdown list when I view this file in github.

File: contributors/design-proposals/envvar-configmap.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+## Design Points
+
+A container can specify a set of existing ConfigMaps to populate environment variables.
+
+There needs to be an easy way to isolate the variables introduced by a given
+ConfigMap. The contents of a ConfigMap may not be known in advance and it may
+be generated by someone or something else. Services may provide binding
+information via a ConfigMap. If you have a common service with multiple
+instances like a Message Queue or Database, there needs to be a way to
+uniquely identify and prevent collision when combining these ConfigMaps.
```
Comment (pmorie) [inline contributors/design-proposals/envvar-configmap.md]: when consuming multiple ConfigMaps in a single pod using this feature.

File: contributors/design-proposals/envvar-configmap.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  // The ConfigMap to select from
+  ConfigMapRef *LocalObjectReference `json:"configMapRef,omitempty"`
+}
+
+type Container struct {
+  // List of sources to populate environment variables in the container.
+  // The keys defined within a source must be a C_IDENTIFIER. An invalid key
+  // will prevent the container from starting. When a key exists in multiple
+  // sources, the value associated with the last source will take precedence.
+  // All env values will take precedence over any listed source.
```
Comment (pmorie) [inline contributors/design-proposals/envvar-configmap.md]: I would clarify `env` here as meaning the container's `env` field.

File: contributors/design-proposals/envvar-configmap.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    ports:
+    - containerPort: 2379
+      protocol: TCP
+    - containerPort: 2380
+      protocol: TCP
+    env:
+    - Name: duplicate_key
+      Value: FROM_ENV
+    - Name: expansion
+      Value: ${REPLACE_ME}
```
Comment (pmorie) [inline contributors/design-proposals/envvar-configmap.md]: The syntax is actually `$(REPLACE_ME)`

**Comment by pmorie** [comment]
After these last nits are fixed, I will merge this.  @fraenkel or @dhoer, please start a new issue / PR for kubectl changes.

**Comment by pmorie** [comment]
Merged, thanks a ton @fraenkel

**Pull request merged**